tooltips: Group tooltips for smooth transition.

shuklamaneesh23 opened the pull request:
This PR groups the tooltips for compose box formatting buttons so that the delay only applies to the first button you hover.

This allows two things:
- Smooth transitions of the tippy between many different reference element targets
- Elements with tooltips next to each other that have a delay can be "grouped" so they appear to share a timeout, which greatly improves UX.
- Previously , the tooltips of formatting buttons were treated independently and each of them had their own delays which doesn't feel smooth while hovering.

Fixes: #24825.
Previous PR Reference: #28387.

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Status  | Behaviour |
|---------|------------------------------------------------------------------------------------------|
| Before  | ![Before](https://github.com/user-attachments/assets/9a1215bc-226b-4597-9588-bdf3776956ca) |
| After   | ![After](https://github.com/user-attachments/assets/55f1850e-d10d-430b-85c3-fda6127a32c4) |





<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shuklamaneesh23 commented:
Hi @alya , this is ready for review.
Thanks!

alya commented:
Thanks! @sayamsamal Adding this your post-release review queue.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
         }
     });
 
+    const compose_buttons = tippy.default(".compose_button_tooltip", {
+        trigger: "mouseenter",
+        placement: "top",
+    });
+
+    tippy.createSingleton(compose_buttons, {
+        onTrigger(instance, event) {
+            const currentTarget = event.currentTarget;
+            if (currentTarget instanceof HTMLElement) {
+                const templateId = currentTarget.dataset.tooltipTemplateId;
+                if (templateId) {
+                    const templateElement = document.querySelector(`#${templateId}`);
+                    if (templateElement) {
+                        const parsedHTML = ui_util.parse_html(templateElement.innerHTML);
+                        instance.setContent(parsedHTML);
+                    }
+                }
+
+                if (currentTarget.classList?.contains("disabled-on-hover")) {
+                    instance.setProps({delay: 0});
+                } else {
+                    instance.setProps({delay: 500});
+                }
+            }
+        },
+    });
+
```
This code doesn't look it belongs in `click_handlers.ts`. If I jog down my memory since that last time I worked on singletons, does it have to do with the instances showing up as empty when used in other places? Say `compose_tooltips.ts`?

sayamsamal replied:
If so, we should still consider defining this logic in `compose_tooltips.ts` and instead initialize the singleton along with it's child instances when the compose box is opened or when the message edit UI is opened.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
+    tippy.createSingleton(compose_buttons, {
+        onTrigger(instance, event) {
+            const currentTarget = event.currentTarget;
+            if (currentTarget instanceof HTMLElement) {
+                const templateId = currentTarget.dataset.tooltipTemplateId;
+                if (templateId) {
+                    const templateElement = document.querySelector(`#${templateId}`);
+                    if (templateElement) {
+                        const parsedHTML = ui_util.parse_html(templateElement.innerHTML);
+                        instance.setContent(parsedHTML);
+                    }
+                }
```
Lets use `get_tooltip_content` from `web/src/tippyjs.ts` instead of redefining an almost identical logic here again.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
+                }
+
+                if (currentTarget.classList?.contains("disabled-on-hover")) {
+                    instance.setProps({delay: 0});
```
Lets revert back to `INSTANT_HOVER_DELAY` defined in `tippyjs.ts`.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
+                if (currentTarget.classList?.contains("disabled-on-hover")) {
+                    instance.setProps({delay: 0});
+                } else {
+                    instance.setProps({delay: 500});
```
Lets revert back to `LONG_HOVER_DELAY` defined in `tippyjs.ts`.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
+    });
+
+    tippy.createSingleton(compose_buttons, {
+        onTrigger(instance, event) {
```
Let's also define the `delay` property for the singletons over here and set it to `delay: LONG_HOVER_DELAY`.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -983,6 +983,33 @@ export function initialize(): void {
+                    instance.setProps({delay: 500});
+                }
+            }
+        },
```
We also need to handle the cleanup for the singletons. Here we would need to destroy the children tippy instances after removing the parent singleton instance.

According to the [docs](https://atomiks.github.io/tippyjs/v6/addons/#destroy):

> When you call singleton.destroy(), the tippyInstances you passed as an argument will not be destroyed also. They are separate instances that can be reused again elsewhere. You should also destroy the tippy instances upon cleanup.

sayamsamal commented:
@shuklamaneesh23 I have left you a list of improvements we can make to the code. From what I remember from the last time I was playing with this, I think we also need to handle the `divider` elements in between the compose controls as they seemed to be interfering with the smoothness advertised on the singletons. Please tag me again after you have made the changes so we can work on that as well.

karlstolley commented on web/templates/compose_control_buttons.hbs:
```diff
@@ -1,66 +1,66 @@
+    <div class="compose_control_button_container preview_mode_disabled compose_button_tooltip" data-tooltip-template-id="add-global-time-tooltip" data-tippy-maxWidth="none">
         <a role="button" class="compose_control_button zulip-icon zulip-icon-time time_pick" aria-label="{{t 'Add global time' }}" tabindex=0></a>
     </div>
     <div class="compose_control_button_container {{#unless giphy_enabled }}hide{{/unless}} preview_mode_disabled" data-tippy-content="{{t 'Add GIF' }}">
```
This is missing the new `compose_button_tooltip` class.

karlstolley commented:
Gave a quick review and tested. I'm of a mind that a visual break in the button bar should correspond with a break in showing tooltips, but I'd be curious for @alya to weigh in on this.

alya commented:
The GIF tooltip seems to be missing?

alya commented:
I think the break feels good in manual testing, but the delay when you go to a new group of buttons does not. I dunno if this is fixable, though; the current implementation seems OK if it's not.

shuklamaneesh23 commented:
Hi @alya , I've updated the PR a little bit.
Could you please test the current behaviour and let me know does it align with what you had in mind?


https://github.com/user-attachments/assets/51f78f57-d25c-41ef-ba27-ad83b2be2152

karlstolley commented:
The one commit here has gotten really massive. I think the singleton tooltip stuff deserves to be its own commit, and probably requires some brief explanatory comments in the code.

sayamsamal commented:
Hey @shuklamaneesh23, I played with the PR in the weekends and I probably think the reason for this disconnect between the tooltip groups is due to the very low values we use for the hide delay for the tippy tooltips. Try setting the delay for the normal options to `[750, 250]` and for the disabled elements setting it as `[100, 250]`. Obviously feel free to play along with those values and experiment how each of them feel.

shuklamaneesh23 commented:
Hi @sayamsamal, thanks for sharing the above information here, that is working quite good. I played by adjusting those delay value to somewhat lower and higher but that ones you shared seems to be perfectly fine.

Also, as `Karl` said, i need to split the commits for adding the singleton tooltip; should i remove the existing tooltip in one commit and add the singleton in another tooltip?

sayamsamal commented on web/src/compose_tooltips.ts:
```diff
@@ -14,10 +14,55 @@ import {pick_empty_narrow_banner} from "./narrow_banner.ts";
+        placement: "top",
+    });
+
+    compose_tooltip_singleton = tippy.createSingleton(compose_tooltips, {
+        delay: LONG_HOVER_DELAY,
+        onTrigger(instance, event) {
+            const currentTarget = event.currentTarget;
+            if (currentTarget instanceof HTMLElement) {
+                const content = get_tooltip_content(currentTarget);
+                if (content) {
+                    instance.setContent(content);
+                }
+                if (currentTarget.classList?.contains("disabled-on-hover")) {
+                    instance.setProps({delay: [100, 250]});
+                } else {
+                    instance.setProps({delay: [750, 250]});
+                }
+            }
+        },
+    });
+}
```
Let's have this block of code be a separate commit, since this is responsible for the actual singleton implementation.

sayamsamal replied:
Let's put the rest of the code in a prep commit prior to this one.

sayamsamal replied:
In the singleton commit, let's define these delay values as constants in `web/src/tippyjs.ts` with appropriate naming such as `SINGLETON_INSTANT_HOVER_DELAY` and `SINGLETON_LONG_HOVER_DELAY` with supporting comments describing why we need these extra delays for the singletons.

shuklamaneesh23 commented:
Hi @sayamsamal, please have a look at it.

**_Updates_**:

**Commit** 1:
Added `compose_button_tooltip` class to all the control buttons.

**Commit** 2:
Removed the old code for instancing tooltips and added new code using createSingleton, as specified.

shuklamaneesh23 commented:
Hi @sayamsamal , made the suggested changes.
Please have another look.
Thanks!

sayamsamal commented:
The changes look good to me, and I'm happy with the current state of this PR. @shuklamaneesh23 can you update the description to either move the screencasts out of the table (they appear broken otherwise), or have GIFs in their place? Also, I'd like to have the Singleton commit be more descriptive, since this will be the first time for us using this feature from the Tippyjs Library, and our fellow contributors would likely want to have more details about what is going on over here. Once you are done with these changes, you can go ahead and mark this PR for integration review, thanks!

timabbott commented on web/src/compose_actions.ts:
```diff
@@ -440,6 +441,7 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
     // compose-box do not cover the last messages of the current stream
     // while writing a long message.
     resize.reset_compose_message_max_height();
+    compose_tooltips.initialize_compose_tooltips();
```
Why is this called from `start`, rather than once from `ui_init`?

timabbott replied:
Well, I see what you're doing architecturally with the save/destroy logic; this works functionally fine for the compose box. 

This PR also completely breaks the tooltips on these elements in the message-edit UI. So you'll need to do some reworking to address that.

sayamsamal replied:
My bad for failing to test the message edit UI detail, I remember having that in my mind (https://github.com/zulip/zulip/pull/33331#discussion_r2031081949) but it got lost somewhere in between.

timabbott commented:
I'm basically satisfied with this pull request except for breaking the message edit UI.

sayamsamal commented on web/src/message_edit.ts:
```diff
@@ -659,6 +659,8 @@ function edit_message($row: JQuery, raw_content: string): void {
         create_copy_to_clipboard_handler($row, the($copy_message), $message_edit_content);
     } else {
         $copy_message.remove();
+        // Initialise the compose tooltips for the edit box.
+        compose_tooltips.initialize_compose_tooltips();
```
@shuklamaneesh23  this call is a duplicate of the next one a little further down. Please rebase the commits to remove this one.

timabbott commented on web/src/message_edit.ts:
```diff
@@ -668,6 +668,10 @@ function edit_message($row: JQuery, raw_content: string): void {
+        compose_tooltips.initialize_compose_tooltips(
+            "edit_message",
+            ".message_edit .compose_button_tooltip",
+        );
```
This doesn't work because you can have multiple edit message widgets open. Look at how `compose_validate` or `upload.ts` handle this issue and you'll see you can fix by including the message ID in your `SingletonContext` type key.

sayamsamal commented on web/src/click_handlers.ts:
```diff
@@ -347,11 +348,17 @@ export function initialize(): void {
     $("body").on("click", ".message_edit_cancel", function (e) {
         const $row = $(this).closest(".message_row");
         message_edit.end_message_row_edit($row);
+        compose_tooltips.clean_up_existing_instances(
+            `edit_message:${$row.attr("data-message-id")}`,
+        );
         e.stopPropagation();
     });
     $("body").on("click", ".message_edit_close", function (e) {
         const $row = $(this).closest(".message_row");
         message_edit.end_message_row_edit($row);
+        compose_tooltips.clean_up_existing_instances(
+            `edit_message:${$row.attr("data-message-id")}`,
+        );
```
Looks like `message_edit.end_message_row_edit()`, is the common call in both the cases, so let's move the cleanup to that method.

sayamsamal commented on web/src/compose_tooltips.ts:
```diff
@@ -14,10 +14,75 @@ import {pick_empty_narrow_banner} from "./narrow_banner.ts";
+
+const compose_button_singleton_context_map = new Map<SingletonContext, SingletonTooltips>();
+
+export function clean_up_existing_instances(context: SingletonContext): void {
```
Let's call this `clean_up_compose_singleton_tooltip` instead, and include a comment as to why we need this function. You can include a reference to http://atomiks.github.io/tippyjs/v6/addons/#destroy.

timabbott commented:
There's still a bug in the state machine here, which I've opened as #35217. Please consider that a priority to work on next.

But the core works, I think that situation is rare enough that we're better off merging this for 11.0, and I don't think that'll be hard to fix as a follow-up in the next few days.

Thanks for all the work on this @shuklamaneesh23 and @sayamsamal!
